File: apps/server/src/api/tabs.rs
use axum::Json;
use axum::extract::{Path, Query, State};
use axum::http::StatusCode;
use serde::Deserialize;
use utoipa::IntoParams;

use crate::api::files::ApiErrorResponse;
use crate::error::ApiError;
use crate::state::AppState;
use crate::tab::{TabInfo, WorktreeTabLayoutState};
pub use crate::tabs::{
    CreateTabRequest, ReorderTabsRequest, UpdateTabRequest, UpdateWorktreeTabLayoutRequest,
};
use crate::util::default_session_id;

#[derive(Debug, Deserialize, IntoParams)]
#[serde(rename_all = "camelCase")]
#[into_params(parameter_in = Query)]
pub struct ListTabsParams {
    #[serde(default = "default_session_id")]
    pub session_id: String,
}

#[utoipa::path(
    get,
    path = "/api/tabs",
    params(ListTabsParams),
    responses(
        (status = 200, description = "List tabs", body = [TabInfo]),
    ),
)]
pub async fn list_tabs(
    State(state): State<AppState>,
    Query(params): Query<ListTabsParams>,
) -> Json<Vec<TabInfo>> {
    Json(state.tabs_service.list_tabs(&params.session_id))
}

#[utoipa::path(
    post,
    path = "/api/tabs",
    request_body = CreateTabRequest,
    responses(
        (status = 201, description = "Tab created", body = TabInfo),
        (status = 400, description = "Invalid tab request", body = ApiErrorResponse),
        (status = 404, description = "Worktree not found", body = ApiErrorResponse),
        (status = 500, description = "Internal server error", body = ApiErrorResponse),
    ),
)]
pub async fn create_tab(
    State(state): State<AppState>,
    Json(req): Json<CreateTabRequest>,
) -> Result<(StatusCode, Json<TabInfo>), ApiError> {
    let (status, tab) = state.tabs_service.create_tab(&state, req).await?;
    Ok((status, Json(tab)))
}

#[utoipa::path(
    delete,
    path = "/api/tabs/{id}",
    params(
        ("id" = String, Path, description = "Tab ID"),
    ),
    responses(
        (status = 204, description = "Tab removed"),
        (status = 404, description = "Tab not found"),
    ),
)]
pub async fn delete_tab(
    State(state): State<AppState>,
    Path(id): Path<String>,
) -> Result<StatusCode, ApiError> {
    close_tab_by_id(&state, &id).await
}

pub(crate) async fn close_tab_by_id(state: &AppState, id: &str) -> Result<StatusCode, ApiError> {
    state.tabs_service.close_tab_by_id(state, id).await
}

#[utoipa::path(
    patch,
    path = "/api/tabs/{id}",
    params(
        ("id" = String, Path, description = "Tab ID"),
    ),
    request_body = UpdateTabRequest,
    responses(
        (status = 200, description = "Tab updated", body = TabInfo),
        (status = 400, description = "Invalid tab update", body = ApiErrorResponse),
        (status = 404, description = "Tab not found"),
    ),
)]
pub async fn update_tab(
    State(state): State<AppState>,
    Path(id): Path<String>,
    Json(req): Json<UpdateTabRequest>,
) -> Result<Json<TabInfo>, ApiError> {
    state
        .tabs_service
        .update_tab(&state, &id, req)
        .await
        .map(Json)
}

#[utoipa::path(
    put,
    path = "/api/projects/{id}/worktrees/{worktreeId}/tab-layout",
    params(
        ("id" = String, Path, description = "Project ID"),
        ("worktreeId" = String, Path, description = "Worktree ID"),
    ),
    request_body = UpdateWorktreeTabLayoutRequest,
    responses(
        (status = 200, description = "Worktree tab layout updated", body = WorktreeTabLayoutState),
        (status = 400, description = "Invalid tab layout", body = ApiErrorResponse),
        (status = 404, description = "Worktree not found", body = ApiErrorResponse),
    ),
)]
pub async fn update_worktree_tab_layout(
    State(state): State<AppState>,
    Path((project_id, worktree_id)): Path<(String, String)>,
    Json(request): Json<UpdateWorktreeTabLayoutRequest>,
) -> Result<Json<WorktreeTabLayoutState>, ApiError> {
    state
        .tabs_service
        .update_worktree_tab_layout(&state, &project_id, &worktree_id, request)
        .await
        .map(Json)
}

#[utoipa::path(
    put,
    path = "/api/tabs/reorder",
    request_body = ReorderTabsRequest,
    responses(
        (status = 200, description = "Tabs reordered", body = [TabInfo]),
        (status = 400, description = "Invalid request"),
    ),
)]
pub async fn reorder_tabs(
    State(state): State<AppState>,
    Json(req): Json<ReorderTabsRequest>,
) -> Result<Json<Vec<TabInfo>>, ApiError> {
    state.tabs_service.reorder_tabs(&state, req).map(Json)
}

File: apps/web/src/lib/api/tabs.ts
import type { components } from "@/lib/contracts/rest.generated";
import type {
  Tab,
  WorktreePaneNode,
  WorktreePaneTabs,
  WorktreeTabLayoutState,
} from "@/lib/types";
import { ApiStatusError, requestJson, requestVoid } from "./client";

type CreateTabRequest = components["schemas"]["CreateTabRequest"];
type UpdateTabRequest = components["schemas"]["UpdateTabRequest"];

export type UpdateWorktreeRestoreStateRequest = {
  activeTabId?: string | null;
  focusedPaneId?: string | null;
  paneMru?: string[] | null;
  tabMruByPane?: Record<string, string[]> | null;
};

export async function listTabs(): Promise<Tab[]> {
  return requestJson("GET", "/api/tabs", {});
}

export async function createTab(request: CreateTabRequest): Promise<Tab> {
  return requestJson("POST", "/api/tabs", { body: request });
}

export async function createTerminalTab(
  worktreeId: string,
  paneId?: string,
): Promise<Tab> {
  return createTab({
    type: "terminal",
    worktreeId: worktreeId,
    paneId: paneId,
  });
}

export async function deleteTab(id: string): Promise<void> {
  try {
    await requestVoid("DELETE", "/api/tabs/{id}", { path: { id } });
  } catch (error) {
    if (!(error instanceof ApiStatusError && error.status === 404)) throw error;
  }
}

export async function updateTab(
  id: string,
  updates: UpdateTabRequest,
): Promise<Tab> {
  return requestJson("PATCH", "/api/tabs/{id}", {
    path: { id },
    body: updates,
  });
}

export async function reorderTabs(
  worktreeId: string,
  paneId: string,
  tabIds: string[],
): Promise<Tab[]> {
  return requestJson("PUT", "/api/tabs/reorder", {
    body: {
      worktreeId: worktreeId,
      paneId: paneId,
      tabIds: tabIds,
    },
  });
}

export async function updateWorktreeTabLayout(
  projectId: string,
  worktreeId: string,
  layout: {
    rootId: string;
    nodes: WorktreePaneNode[];
    panes: WorktreePaneTabs[];
  },
): Promise<WorktreeTabLayoutState> {
  return requestJson(
    "PUT",
    "/api/projects/{id}/worktrees/{worktreeId}/tab-layout",
    {
      path: { id: projectId, worktreeId: worktreeId },
      body: {
        rootId: layout.rootId,
        nodes: layout.nodes,
        panes: layout.panes,
      },
    },
  );
}

export async function updateWorktreeRestoreState(
  projectId: string,
  worktreeId: string,
  restoreState: UpdateWorktreeRestoreStateRequest,
): Promise<void> {
  try {
    await requestVoid(
      "PUT",
      "/api/projects/{id}/worktrees/{worktreeId}/restore-state",
      {
        path: { id: projectId, worktreeId: worktreeId },
        body: {
          activeTabId: restoreState.activeTabId ?? null,
          focusedPaneId: restoreState.focusedPaneId ?? null,
          paneMru: restoreState.paneMru ?? [],
          tabMruByPane: restoreState.tabMruByPane ?? {},
        },
      },
    );
  } catch (error) {
    if (!(error instanceof ApiStatusError && error.status === 404)) throw error;
  }
}
